[webkitcorepy] Create process-like Thread object

JonWBedard opened the pull request:
#### dc2cb0fa4c7ff63091e623a7c48757a3db47a006
<pre>
[webkitcorepy] Create process-like Thread object
<a href="https://bugs.webkit.org/show_bug.cgi?id=245913">https://bugs.webkit.org/show_bug.cgi?id=245913</a>
&lt;rdar://100642035&gt;

Reviewed by Dewei Zhu and David Kilzer.

* Tools/Scripts/libraries/webkitcorepy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py: Import Thread object
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/subprocess_utils.py:
(Thread):
(Thread.terminated): Check if the current thread has it&apos;s termination flag flipped.
(Thread.__init__):
(Thread.poll): Check if the Thread is running.
(Thread.terminate): Set terminate flag.
(Thread.kill): Ditto.
(Thread.__enter__): Start thread.
(Thread.__exit__): Join thread.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/subprocess_utils_unittest.py:
(SubprocessUtils.test_thread):
(SubprocessUtils.test_killed_thread):

Canonical link: <a href="https://commits.webkit.org/255370@main">https://commits.webkit.org/255370@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/abee67231725d211cca473174baeb1033aa2d698

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92316 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1546 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/22902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/102089 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96317 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1542 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/84744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/98253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97977 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/78831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/84744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/95889 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/82934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/22902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/84744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/36348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/22902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/91363 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/34111 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/22902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/37984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/78831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1680 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39885 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/22902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

ddkilzer reviewed (approved):
r=me

dewei-zhu commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/subprocess_utils.py:
```diff
@@ -146,3 +147,32 @@ def decode(string):
+        return None if self.is_alive() else {True: 1, False: 0}.get(self._terminated, -1)
+
+    def terminate(self):
+        self._terminated = True
```
Would this be sufficient to kill the thread?

JonWBedard replied:
No.

There isn't really a platform agnostic way to kill a thread. This code relies on the the thread being "killed" looking this flag (which is exactly what the unit tests do). I thought that was the best way to match the API without creating something complicated, which ultimately, we don't need (all we need is a way for the parent thread to stop an infinite loop its worker threads)

dewei-zhu reviewed (approved):
r=me

JonWBedard committed:
[webkitcorepy] Create process-like Thread object
https://bugs.webkit.org/show_bug.cgi?id=245913
<rdar://100642035>

Reviewed by Dewei Zhu and David Kilzer.

* Tools/Scripts/libraries/webkitcorepy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py: Import Thread object
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/subprocess_utils.py:
(Thread):
(Thread.terminated): Check if the current thread has it's termination flag flipped.
(Thread.__init__):
(Thread.poll): Check if the Thread is running.
(Thread.terminate): Set terminate flag.
(Thread.kill): Ditto.
(Thread.__enter__): Start thread.
(Thread.__exit__): Join thread.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/subprocess_utils_unittest.py:
(SubprocessUtils.test_thread):
(SubprocessUtils.test_killed_thread):

Canonical link: https://commits.webkit.org/255370@main
